### PR TITLE
Feature/review most retried tests and do workaround

### PR DESCRIPTION
**Pull request by asanzgom**
closes https://issues.redhat.com/browse/MGDSTRM-10622

### CODE REVIEW

**Comment by kornys** [comment]
@asanzgom could you please elaborate about timeouts you set?

File: lib/topic.ts [hunk 1]
```diff
@@ -3,11 +3,11 @@ import { showElementActions } from './kafka';
 import { navigateToKafkaList } from './navigation';
 
 export const navigateToKafkaTopicsList = async function (page: Page, kafkaName: string) {
-  await expect(page.getByText(kafkaName)).toHaveCount(1);
-  await page.locator('a', { hasText: kafkaName }).click();
-  await expect(await page.locator('button[aria-label="Topics"]')).toHaveCount(1);
+  await expect(page.getByText(`${kafkaName}`)).toHaveCount(1);
+  await page.locator('a', { hasText: `${kafkaName}` }).click();
```
Comment (Frawless) [inline lib/topic.ts:7]: Why this change? Is there some difference in variable evaluation? I didn't find it in pw doc

File: lib/topic.ts [hunk 2]
```diff
@@ -3,11 +3,11 @@ import { showElementActions } from './kafka';
...
+  await expect(page.locator('button[aria-label="Topics"]')).toHaveCount(1, { timeout: 10000 });
   // data-testid=pageKafka-tabTopics
-  await page.locator('button[aria-label="Topics"]').click();
+  await page.locator('button[aria-label="Topics"]').click({ timeout: 10000 });
```
Comment (Frawless) [inline lib/topic.ts:10]: Why do we decrease it from 30 to 10s? Or is the default value different?

File: lib/topic.ts [hunk 3]
```diff
@@ -73,16 +73,16 @@ export const deleteKafkaTopic = async function (page: Page, name: string) {
...
-  await expect(await page.locator('button', { hasText: 'Messages' })).toHaveCount(1);
+  await expect(page.locator('button', { hasText: 'Messages' })).toHaveCount(1);
   await page.locator('button', { hasText: 'Messages' }).click();
 };
 
 export const navigateToProperties = async function (page: Page, kafkaName: string, topicName: string) {
   await navigateToKafkaList(page);
   await navigateToKafkaTopicsList(page, kafkaName);
-  await expect(await page.locator('a', { hasText: `${topicName}` })).toHaveCount(1);
+  await expect(page.locator('a', { hasText: `${topicName}` })).toHaveCount(1, { timeout: 6000 });
```
[the remark this answers is not in the document]
Comment (asanzgom) [reply lib/topic.ts:85]: Is it really waiting 30 seconds? Why does it fail then with 30 seconds and not with 6?

File: lib/topic.ts (hunk 1 above)
Comment (asanzgom) [reply lib/topic.ts:7]: It is to tighten that the String value of the variable is used, to make sure

File: lib/topic.ts (hunk 2 above)
Comment (asanzgom) [reply lib/topic.ts:10]: by allowing 10 seconds on those steps that were the ones to repeatedly fail we are allowing time for the buttons to load.

**Comment by asanzgom** [comment]
> @asanzgom could you please elaborate about timeouts you set?

sure, replied above

File: lib/topic.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Frawless) [reply lib/topic.ts:85]: Well, it should, see https://github.com/redhat-developer/consoledot-e2e/blob/main/playwright.config.ts#L23 ... maybe we interpret it in the wrong way?

File: lib/topic.ts (hunk 1 above)
Comment (Frawless) [reply lib/topic.ts:7]: `kafkaName` has type `string`, can you elaborate more when it wasn't string?

File: lib/topic.ts (hunk 2 above)
Comment (Frawless) [reply lib/topic.ts:10]: Oh ye, we don't have timeout for clicking. Anyway line 8 waits for the button with timeout 30s so when PW click on it, it should be here for sure. I assume it doesn't work for you, why?

File: lib/topic.ts (hunk 1 above)
Comment (asanzgom) [reply lib/topic.ts:7]: both ways resolve to a String, maybe this one is more robust. The same `${}` is used throughout the project @mmusil Can you confirm, please?

File: lib/topic.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (asanzgom) [reply lib/topic.ts:85]: that is https://playwright.dev/docs/test-timeouts#expect-timeout Assertion timeout is unrelated to the test timeout. There are several different kinds of Timeout, this is acting on the lines of code that were repeatedly failing

[the remark this answers is not in the document]
Comment (Frawless) [reply lib/topic.ts:85]: But you are setting timeout for expect to 6s but our default value is 30s. This is not related to test timeout that is set to 20 minutes.

File: lib/topic.ts (hunk 1 above)
Comment (Frawless) [reply lib/topic.ts:7]: Where it is used? Could you share some docs where the robustness is described?

File: lib/topic.ts (hunk 2 above)
Comment (asanzgom) [reply lib/topic.ts:10]: well, the traces when it failed showed the Button was not there, and therefore could not be clicked. With this Timeout I got 2 sucessful executions in a row, and it has already passed PR Pipeline

File: lib/topic.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (asanzgom) [reply lib/topic.ts:85]: Expect timeout
Web-first assertions like expect(locator).toHaveText() have a separate timeout, 5 seconds by default. Assertion timeout is unrelated to the test timeout.

File: lib/topic.ts (hunk 2 above)
Comment (Frawless) [reply lib/topic.ts:10]: I think we should understand why the current approach doesn't work. I am not against adding the timeout, but line 8 should have 30s now, you decrease it to 10s and you add an additional timeout to line 10 that should pass immediately because of line 8.

File: lib/topic.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Frawless) [reply lib/topic.ts:85]: https://github.com/redhat-developer/consoledot-e2e/blob/main/playwright.config.ts#L18

```
export default defineConfig({
  expect: {
    timeout: 10 * 1000,
  },
});
```

I don't think there is a difference between doc and our usage

File: lib/topic.ts (hunk 2 above)
Comment (asanzgom) [reply lib/topic.ts:10]: Can I resolve this conversation?

File: lib/topic.ts (hunk 1 above)
Comment (asanzgom) [reply lib/topic.ts:7]: Can I resolve this conversation?

File: lib/topic.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (asanzgom) [reply lib/topic.ts:85]: Can I resolve this conversation?

File: lib/topic.ts (hunk 1 above)
Comment (Frawless) [reply lib/topic.ts:7]: I still don't see any prove that your new code is more robust than the old one

File: lib/topic.ts (hunk 2 above)
Comment (Frawless) [reply lib/topic.ts:10]: which kind of problems? The code should behave as expected especially when we wait for element X for 30s, in case that expect does not fail it means the X is available. Then we click on X and this fails - WHY?

File: lib/topic.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Frawless) [reply lib/topic.ts:85]: No, because I still didn't get a proper explanation of why this is needed.

File: lib/topic.ts (hunk 1 above)
Comment (asanzgom) [reply lib/topic.ts:7]: As I mentioned above the same way it is being used throughout the project, I triggered tests, they are passing twice in a row: 
image
image

File: lib/topic.ts (hunk 2 above)
Comment (asanzgom) [reply lib/topic.ts:10]: the failures that lead to retrials. I am not really sure the test behaves like that but forcing the assertion to happen for these specified 10 seconds

File: lib/topic.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Frawless) [reply lib/topic.ts:85]: do you understand what I am trying to say by all of the lines above? Is there some bug in pw that did not respect the configuration from the config file? Or why does changing the timeout form 30s to 6s solve it? Or it is somehow related to removing await - if so, why? The expected timeout should work fine anyway.

[the remark this answers is not in the document]
Comment (asanzgom) [reply lib/topic.ts:85]: Ok, so do you suggest I should only remove the unnecessary await and don't add the timeout?

File: lib/topic.ts (hunk 1 above)
Comment (Frawless) [reply lib/topic.ts:7]: This still doesn't explain the need for this change. We should follow one pattern. Anyway, I guess I can live with it...

File: lib/topic.ts (hunk 2 above)
Comment (ldimaggi) [reply lib/topic.ts:10]: Sorry to be late to this thread. The conversation is confusing. 

I agree that we want to avoid "magic numbers" - if the default timeout is 30 seconds, are you saying that reducing this specific timeout to 10 seconds is resulting in successful tests executions?

Comment (asanzgom) [reply lib/topic.ts:10]: No, I am not sure that at that step the timeout of 30 seconds was happening

Comment (mmusil) [reply lib/topic.ts:10]: This does not seem right @asanzgom, could you please try a negative test to verify that the PW config for 30 sec expect timeout is truly not applied and therefore there is a bug in Playwright? And upload a trace here that would be showing `expect timeout reached 30 sec` or `expect timeout reached 5 sec` or whatever?

Comment (Frawless) [reply lib/topic.ts:10]: @mmusil good idea, thanks

File: lib/topic.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Frawless) [reply lib/topic.ts:85]: You can try it, all I am saying is, that your change basically doesn't change how the code should behave so it will probably don't fix the real problem. Unless we don't fully understand how pw should work, this might be true as well :) But I didn't get any proper explanation why your changes are working and the current code not

File: lib/topic.ts (hunk 2 above)
Comment (ldimaggi) [reply lib/topic.ts:10]: And - on line 8 - we have a default expect timeout of 30, and the change reduces this to 10?

**Comment by asanzgom** [comment]
We got to the solution for this that is to await expect for the Consumer Groups page to load, which is what happens after saving changes on the Edit Topic Properties menu and should be awaited otherwise some actions are being taken during the 'Save' callback and therefore creates asynchronicity sometimes that leads to not finding the expected Topics Tab button. 

I close this Pull Request and open a new one with the solution that we have tested to work
